### PR TITLE
Add text verification logic when editing descriptions/captions.

### PR DESCRIPTION
**Pull request by dbrant**
https://phabricator.wikimedia.org/T256816

### CODE REVIEW

File: app/src/main/java/org/wikipedia/descriptions/DescriptionEditView.java [hunk 1]
```diff
@@ -249,10 +250,38 @@ public boolean showingReviewContent() {
     }
 
     public void setError(@Nullable CharSequence text) {
+        pageDescriptionLayout.setErrorIconDrawable(R.drawable.ic_error_black_24dp);
+        pageDescriptionLayout.setErrorIconTintList(ColorStateList.valueOf(ResourceUtil.getThemedColor(getContext(), R.attr.colorError)));
+        pageDescriptionLayout.setErrorTextColor(ColorStateList.valueOf(ResourceUtil.getThemedColor(getContext(), R.attr.colorError)));
```
Comment (cooltey) [inline app/src/main/java/org/wikipedia/descriptions/DescriptionEditView.java:255]: By looking at the Zeplin, should the error text to be bolded?

File: app/src/main/java/org/wikipedia/descriptions/DescriptionEditView.java [hunk 2]
```diff
@@ -249,10 +250,38 @@ public boolean showingReviewContent() {
...
+        layoutErrorState(text);
+    }
+
+    private void setWarning(@Nullable CharSequence text) {
+        pageDescriptionLayout.setErrorIconDrawable(R.drawable.ic_warning_24);
+        pageDescriptionLayout.setErrorIconTintList(ColorStateList.valueOf(ContextCompat.getColor(getContext(), R.color.yellow30)));
+        pageDescriptionLayout.setErrorTextColor(ColorStateList.valueOf(ContextCompat.getColor(getContext(), R.color.yellow30)));
```
Comment (cooltey) [inline app/src/main/java/org/wikipedia/descriptions/DescriptionEditView.java:262]: By looking at the Zeplin, should the error text to be bolded?

**Comment by dbrant** [comment]
> Found one issue:
> When it's showing the error message, the voice input icon will appear for a short time after rotating the device; and if it is showing a warning message, it will turn the color into the red for a short time after rotating the device.

I think your other PR takes care of this automatically.
Regarding the error text, I asked Robin on the ticket. Can be a follow-up PR.

**Comment by dbrant** [comment]
> One last thing regarding rotating the device:
> If it is showing the warning message (yellow), when rotating the device, the text color will go red and then change to yellow.

All done, including bold error text.

File: app/src/main/java/org/wikipedia/analytics/ABTestDescriptionEditChecksFunnel.java [hunk 3]
```diff
@@ -0,0 +1,20 @@
+package org.wikipedia.analytics;
+
+public final class ABTestDescriptionEditChecksFunnel extends ABTestFunnel {
+    public ABTestDescriptionEditChecksFunnel() {
+        super("descriptionEditChecks", ABTestFunnel.GROUP_SIZE_2);
+    }
+
+    public boolean shouldSeeChecks() {
+        boolean enrolled = isEnrolled();
+        boolean showChecks = getABTestGroup() == ABTestFunnel.GROUP_1;
```
Comment (cooltey) [inline app/src/main/java/org/wikipedia/analytics/ABTestDescriptionEditChecksFunnel.java:10]: Would it be possible to add `ReleaseUtil.isPreBetaRelease()` for testing purposes?

File: app/src/main/java/org/wikipedia/descriptions/DescriptionEditView.java [hunk 4]
```diff
@@ -292,8 +319,41 @@ private void performReadArticleClick() {
...
+            removeCallbacks(textValidateRunnable);
+            postDelayed(textValidateRunnable, TEXT_VALIDATE_DELAY_MILLIS);
+        } else {
+            isTextValid = true;
+            updateSaveButtonEnabled();
+            setError(null);
+        }
+    }
+
+    void validateText() {
```
Comment (cooltey) [inline app/src/main/java/org/wikipedia/descriptions/DescriptionEditView.java:332]: Should we remove the spaces at the beginning and the end of the text? Users can still send the description if they add a space to the end even if the text contains an invalid format.

File: app/src/main/java/org/wikipedia/analytics/ABTestDescriptionEditChecksFunnel.java (hunk 3 above)
Comment (dbrant) [reply app/src/main/java/org/wikipedia/analytics/ABTestDescriptionEditChecksFunnel.java:16]: Yep

File: app/src/main/java/org/wikipedia/descriptions/DescriptionEditView.java (hunk 4 above)
Comment (dbrant) [reply app/src/main/java/org/wikipedia/descriptions/DescriptionEditView.java:332]: Sensible.

**Pull request merged**